csharp/src/Ice/OutgoingResponseFrame.cs

```diff
@@ -22,27 +22,25 @@ public sealed class OutgoingResponseFrame : OutputStream
         /// <summary>The response context. Always null with Ice1.</summary>
         public Context? Context { get; }
 
-        /// <summary>Creates a new "success" outgoing response frame.</summary>
+        private EncodingVersion _payloadEncoding; // TODO: move to OutputStream
+
+        /// <summary>Creates a new outgoing request frame with an OK reply status and a void return value.</summary>
         /// <param name="current">The current parameter holds decoded header data and other information about the
         /// request for which this method creates a response.</param>
-        /// <param name="format">The Slice format (Compact or Sliced) used by the encapsulation.</param>
-        /// <param name="payloadWriter">An action that writes the contents of the payload.</param>
-        public OutgoingResponseFrame(Current current, FormatType? format, Action<OutputStream> payloadWriter)
-            : this(current.Adapter.Communicator, current.RequestId)
-        {
-            WriteByte((byte)ReplyStatus.OK);
-            StartEncapsulation(current.Encoding, format);
-            payloadWriter(this);
-            EndEncapsulation();
-        }
+        public static OutgoingResponseFrame Empty(Current current)
+            => new OutgoingResponseFrame(current, ReplyStatus.OK, ArraySegment<byte>.Empty);
 
-        /// <summary>Creates a new "success" outgoing response frame with a null format</summary>
+        /// <summary>Creates a new outgoing response frame. It's a partial frame, and its payload needs to be
+        /// written later on with StartReturnValue/EndReturnValue.</summary>
         /// <param name="current">The current parameter holds decoded header data and other information about the
         /// request for which this method creates a response.</param>
-        /// <param name="payloadWriter">An action that writes the contents of the payload.</param>
-        public OutgoingResponseFrame(Current current, Action<OutputStream> payloadWriter)
-            : this(current, format: null, payloadWriter)
+        public OutgoingResponseFrame(Current current) : base(current.Adapter.Communicator, Util.CurrentProtocolEncoding)
         {
+            RequestId = current.RequestId;
+            _payloadEncoding = current.Encoding;
```

Comment:
The encoding of the frame's header depends entirely on the protocol: ice1 => encoding 1.0 (or 1.1 - they are identical for the subset of the Ice 1.x encoding used by the ice1 headers), ice2 => encoding 2.0. "CurrentProtocolEncoding" is currently 1.0 ... we need to clean this up since Ice 4.0 will not include support for the 1.0 encoding (which is btw contrary to our old encoding version rules).
The encoding of the payload can be anything and does not depend on the protocol. Here, we use current.Encoding because the response for a request must use the same encoding for its result as the in-parameters of the request (which is what current.Encoding represents).